### PR TITLE
executor: fix revoke result view bug   

### LINKED ISSUE #38421: User information exists after the privileges are revoked in `mysql`.`tables_priv`
## Bug Report

Thanks for watching the issue. I found a MySQL compatibility broken case, 
which is the `mysql`.`tables` have rows that user don't have any privileges to the table.

### 1. Minimal reproduce step (Required)

```
CREATE user test;
GRANT SELECT ON `mysql`.`db` TO test;
REVOKE SELECT ON `mysql`.`db` FROM test;
SELECT * FROM `mysql`.`tables_priv`;
```

### 2. What did you expect to see? (Required)

`mysql`.`tables_priv` doesn't have the row related to `test` user.

### 3. What did you see instead (Required)

I assumed that what `Table_priv` is `NULL` means the `test` user don't have any privileges to the table.
But, I think it's better to remove the row.

```
MySQL [(none)]> SELECT * FROM `mysql`.`tables_priv`;
+------+-------+------+------------+---------+---------------------+------------+-------------+
| Host | DB    | User | Table_name | Grantor | Timestamp           | Table_priv | Column_priv |
+------+-------+------+------------+---------+---------------------+------------+-------------+
| %    | mysql | test | db         |         | 2022-10-12 15:56:22 |            |             |
+------+-------+------+------------+---------+---------------------+------------+-------------+
1 row in set (0.03 sec)
```

When I test with MySQL 5.7, it removes the line when the user privilege is revoked.

### 4. What is your TiDB version? (Required)

```
MySQL [(none)]> SELECT tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                       |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v6.1.1
Edition: Community
Git Commit Hash: 5263a0abda61f102122735049fd0dfadc7b7f8b2
Git Branch: heads/refs/tags/v6.1.1
UTC Build Time: 2022-08-25 10:42:41
GoVersion: go1.18.5
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)
```

### PR DESCRIPTION
**Pull request by wxbty**
Signed-off-by: x-shadow-man <1494445739@qq.com>

Fix bug in issue #38421
also on column-level privilege 

### What problem does this PR solve?

This pr mainly fixes the bug of 38421.
To this end, I added  new unit test【TestRevokeTableSingle、TestRevokeTableSingleColumn】, and also modified the involved unit tests【TestRevokeTableScope、TestRevokeColumnScope】.

Issue Number: close #38421

### CODE REVIEW

**Comment by wxbty** [comment]
@cfzjywxk @you06   Do you have time? Please take a look,  thanks!

**Comment by wxbty** [comment]
hi, @dveeden   This pr also wants to get your advice,thx

**Comment by wxbty** [comment]
I noticed the CI test, maybe it was affected, if this is not a bug, I will close the pr

File: executor/revoke.go [hunk 1]
```diff
@@ -271,7 +271,16 @@ func (e *RevokeExec) revokeTablePriv(internalSession sessionctx.Context, priv *a
...
 	sql := new(strings.Builder)
 	sqlexec.MustFormatSQL(sql, "UPDATE %n.%n SET ", mysql.SystemDB, mysql.TablePrivTable)
-	err = composeTablePrivUpdateForRevoke(internalSession, sql, priv.Priv, user, host, dbName, tblName)
+	isDelRow, err := composeTablePrivUpdateForRevoke(internalSession, sql, priv.Priv, user, host, dbName, tblName)
+	if err != nil {
+		return err
+	}
+
+	if isDelRow {
+		delSQL := new(strings.Builder)
```
Comment (xhebox) [inline executor/revoke.go:280]: You can reuse `strings.Builder` by `sql.Reset()`

File: executor/revoke_test.go [hunk 2]
```diff
@@ -166,6 +172,21 @@ func TestRevokeColumnScope(t *testing.T) {
 	tk.MustQuery(`SELECT Column_priv FROM mysql.Columns_priv WHERE User="testCol1Revoke" and host="localhost" and db="test" and Table_name="test3"`).Check(testkit.Rows(""))
 }
 
+func TestRevokeTableSingle(t *testing.T) {
```
Comment (xhebox) [inline executor/revoke_test.go:175]: Add a comment like  `ref #38421` to link the original bug.

**Review by xhebox — COMMENTED** [review]
LGTM otherwise

**Review by xhebox — APPROVED** [review]

**Review by you06 — COMMENTED** [review]
LGTM
Can you help fix a similar issue on column-level privilege?

```sql
CREATE user test;
GRANT SELECT(`Host`) ON `mysql`.`db` TO test;
REVOKE SELECT(`Host`) ON `mysql`.`db` FROM test;
SELECT * FROM `mysql`.`columns_priv`; -- Expected empty result set, but got the following rows
+------+-------+------+------------+-------------+---------------------+-------------+
| Host | DB    | User | Table_name | Column_name | Timestamp           | Column_priv |
+------+-------+------+------------+-------------+---------------------+-------------+
| %    | mysql | test | db         | Host        | 2022-10-26 16:01:23 |             |
+------+-------+------+------------+-------------+---------------------+-------------+
1 row in set (0.002 sec)

```

**Comment by wxbty** [comment]
> LGTM Can you help fix a similar issue on column-level privilege?
> 
> ```sql
> CREATE user test;
> GRANT SELECT(`Host`) ON `mysql`.`db` TO test;
> REVOKE SELECT(`Host`) ON `mysql`.`db` FROM test;
> SELECT * FROM `mysql`.`columns_priv`; -- Expected empty result set, but got the following rows
> +------+-------+------+------------+-------------+---------------------+-------------+
> | Host | DB    | User | Table_name | Column_name | Timestamp           | Column_priv |
> +------+-------+------+------------+-------------+---------------------+-------------+
> | %    | mysql | test | db         | Host        | 2022-10-26 16:01:23 |             |
> +------+-------+------+------------+-------------+---------------------+-------------+
> 1 row in set (0.002 sec)
> ```

ok

**Review by dveeden — APPROVED** [review]

**Comment by wxbty** [comment]
> LGTM Can you help fix a similar issue on column-level privilege?
> 
> ```sql
> CREATE user test;
> GRANT SELECT(`Host`) ON `mysql`.`db` TO test;
> REVOKE SELECT(`Host`) ON `mysql`.`db` FROM test;
> SELECT * FROM `mysql`.`columns_priv`; -- Expected empty result set, but got the following rows
> +------+-------+------+------------+-------------+---------------------+-------------+
> | Host | DB    | User | Table_name | Column_name | Timestamp           | Column_priv |
> +------+-------+------+------------+-------------+---------------------+-------------+
> | %    | mysql | test | db         | Host        | 2022-10-26 16:01:23 |             |
> +------+-------+------+------------+-------------+---------------------+-------------+
> 1 row in set (0.002 sec)
> ```

@xhebox  PTAL

File: executor/revoke.go [hunk 3]
```diff
@@ -306,6 +315,17 @@ func (e *RevokeExec) revokeColumnPriv(internalSession sessionctx.Context, priv *
 		if err != nil {
 			return err
 		}
+
+		if isDelRow {
+			sql.Reset()
+			sqlexec.MustFormatSQL(sql, "DELETE FROM %n.%n WHERE User=%? AND Host=%? AND DB=%? AND Table_name=%?", mysql.SystemDB, mysql.ColumnPrivTable, user, host, dbName, tbl.Meta().Name.O)
```
Comment (you06) [inline executor/revoke.go:321]: Add `Column_name = %?` condition.

**Review by you06 — APPROVED** [review]

**Comment by wxbty** [comment]
> The branch is out-of-date, because the PR is from [x-shadow-man](https://github.com/x-shadow-man)'s repo), I cannot update branch for you, please click the update branch button, see [help doc](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/keeping-your-pull-request-in-sync-with-the-base-branch#updating-your-pull-request-branch).

done

**Comment by xhebox** [comment]
> > The branch is out-of-date, because the PR is from [x-shadow-man](https://github.com/x-shadow-man)'s repo), I cannot update branch for you, please click the update branch button, see [help doc](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/keeping-your-pull-request-in-sync-with-the-base-branch#updating-your-pull-request-branch).
> 
> done

Okay, so you need to update it again. I have no idea why I can't do that. Maybe you have disabled `allow maintainers to edti`.

**Comment by wxbty** [comment]
> 

done

**Comment by wxbty** [comment]
> /merge



> /merge

please try again

**Comment by xhebox** [comment]
> > /merge
> 
> > /merge
> 
> please try again

Okay, again.. Maybe you can fork a new repo and file the pull request by yourself. Or wait the night: less PRs will be merged.

**Comment by wxbty** [comment]
> > > /merge
> > 
> > 
> > > /merge
> > 
> > 
> > please try again
> 
> Okay, again.. Maybe you can fork a new repo and file the pull request by yourself. Or wait the night: less PRs will be merged.


I have the permission of the x-shadow man repository, but I can't find the modification of the allow maintainers to edti option. I think it is only available when pr is just opened.
Do I need to reopen a pr? Or try again?

**Comment by xhebox** [comment]
> 
> I have the permission of the x-shadow man repository, but I can't find the modification of the allow maintainers to edti option.

The button should be somewhere at the right side panel.

> I think it is only available when pr is just opened. Do I need to reopen a pr? Or try again?

It is up to you. I suggest opening another PR with your  repo will be easy. I'll merge the PR when I am notified. 

> Then Why does the merge operation take so long? until timeout

CI is always unstable somehow. And it does take a long time to run all tests.

**Comment by wxbty** [comment]
> /merge

When ci is passed, maybe you will merge successfully.

**Comment by xhebox** [comment]
> > /merge
> 
> When ci is passed, maybe you will merge successfully.

Yes, so I tag it as mergeable first. So I don't need to watch the PR to see tests are passed or not.

**Comment by wxbty** [comment]
> > > /merge
> > 
> > 
> > When ci is passed, maybe you will merge successfully.
> 
> Yes, so I tag it as mergeable first. So I don't need to watch the PR to see tests are passed or not.

If this time unsuccessful, I will reopen it, I know how the new pr is set up

**Comment by wxbty** [comment]
> /merhe

typo

**Pull request merged**